### PR TITLE
Preserve Permissions on Zipfile extraction

### PR DESCRIPTION
**Pull request by johnwparent**
#24556 merged in support for Python's .zip file support via `ZipFile`, however as per #30200 `ZipFile` does not preserve file permissions of the extracted contents. This PR subclasses `ZipFile` to allow the propagation of permissions.

### CODE REVIEW

File: lib/spack/spack/util/compression.py [hunk 1]
```diff
@@ -39,6 +40,39 @@ def _gunzip(archive_file):
...
 
 
+def spack_unzipper():
+    from zipfile import ZipFile, ZipInfo
+
+    # implementation based on https://stackoverflow.com/a/39296577
+    class SpackZipFile(ZipFile):
+        if sys.version_info < (3, 6):
+            def extract(self, member, path=None, pwd=None):
+                # below is a direct copy of ZipFile's extract method
```
Comment (scheibelp) [inline lib/spack/spack/util/compression.py]: I don't understand: if this is an exact copy: why is it redefined here? Why not:

```
if sys.version_info >= (3, 7):
  def _extract_member(...):
    ...
# no else statement needed
```

File: lib/spack/spack/util/compression.py [hunk 2]
```diff
@@ -39,6 +40,39 @@ def _gunzip(archive_file):
...
+        else:
+            def _extract_member(self, member, targetpath, pwd):
+                # Call ZipFile extract member method
+                if not isinstance(member, ZipInfo):
+                    member = self.getinfo(member)
+                path = super(ZipFile, self)._extract_member(member, targetpath, pwd)
+
+                # Catch output and restore permissions if permissions attr is non zero
+                if member.external_attr > 0xffff:
+                    os.chmod(path, member.external_attr >> 16)
```
Comment (scheibelp) [inline lib/spack/spack/util/compression.py]: Normally I'd ask to have this documented, but according to https://bugs.python.org/issue18262 there is no documentation for this attribute (I do think that link would be useful in a comment here and possibly also https://bugs.python.org/issue15795, since that would cover the reason why there is fiddling with `external_attr` at all).

I'm therefore inclined to avoid using `zipfile` whenever possible (i.e. a hybrid of this and https://github.com/spack/spack/pull/30200 - do this stuff on Windows but don't use `zipfile` on other systems).

File: lib/spack/spack/util/compression.py [hunk 3]
```diff
@@ -49,9 +83,9 @@ def _unzip(archive_file):
         archive_file (str): absolute path of the file to be decompressed
     """
     try:
-        from zipfile import ZipFile
+        SpackZip = spack_unzipper()
         destination_abspath = os.getcwd()
-        with ZipFile(archive_file, 'r') as zf:
+        with SpackZip(archive_file, 'r') as zf:
```
Comment (scheibelp) [inline lib/spack/spack/util/compression.py]: I missed it in earlier review but there is a conceptual mismatch in terms of the return value of the function: we either return a `callable` object, or we extract the archive file and return `None`, we should be consistent about that.

**Review by scheibelp — CHANGES_REQUESTED** [review]
Based on the lack of documentation in `zipfile` related to these changes, let alone an official API, I think we should use `zipfile` only on Windows (like in https://github.com/spack/spack/pull/30200 - but the changes here are still needed when running on Windows). I have more details in the comments. Let me know if you disagree.

**Comment by scheibelp** [comment]
Other related links:

* https://stackoverflow.com/questions/434641/how-do-i-set-permissions-attributes-on-a-file-in-a-zip-file-using-pythons-zip
* https://stackoverflow.com/questions/279945/set-permissions-on-a-compressed-file-in-python

**Comment by scheibelp** [comment]
Regarding https://github.com/spack/spack/pull/30407#issuecomment-1124222747, these are only loosely related I think: they are about setting permissions inside of the `zipfile`, which seems strange given that this whole PR is about how those permissions aren't respected at all.

File: lib/spack/spack/util/compression.py (hunk 1 above)
Comment (johnwparent) [reply lib/spack/spack/util/compression.py]: That comment is a mistake that I left from a first pass and missed on cleanup. I do agree that we only need to override `_extract_member`, but I was confused as to the API changes per version, I just went through the source and it looks like as long as we override `_extract_member` the version will not matter.

File: lib/spack/spack/util/compression.py (hunk 3 above)
Comment (johnwparent) [reply lib/spack/spack/util/compression.py]: Agreed. I'll have this return None on all platforms as a callable ZipFile would require passing a function pointer to `ZipFile.extractall`, and I'd rather keep control over that process on the compression side. If the convention is a callable, that can work as well though. 

Note: the compression PR overhauls a large amount of this, and would be useful to have in this scenario.

Comment (johnwparent) [reply lib/spack/spack/util/compression.py]: I'm surprised systems without `ZipFile` available weren't failing on the returned unzip callable.

File: lib/spack/spack/util/compression.py (hunk 2 above)
Comment (johnwparent) [reply lib/spack/spack/util/compression.py]: I did not find that issue, but also struggled to find documentation, glad to know it wasn't because I was bad at searching. 

Is the hesitancy around using an undocumented variable? I'd rather avoid as much separation of the Windows/other platform interfaces as possible, and it seems like the issue would still exist on Windows as well, so IMO, we may as well develop an alternative solution that works for both platforms.
Creating separate interfaces for different platforms feels like an easy way to end up maintaining two separate codebases that exist on top of each other. Obviously there will be some unavoidable instances where we need to have truly separate interfaces (i.e. the `posixpath` vs `nt` modules) and this may be once of those scenarios, but I think putting some extra effort into working toward a unified solution would be worth it.
If you feel differently, I'll happily implement in the manner suggested in the above comment (and that may be what we need to do anyway if we can't determine a better method)

Comment (scheibelp) [reply lib/spack/spack/util/compression.py]: > Is the hesitancy around using an undocumented variable?

Yes: if they don't document it, then they can change it whenever they want. Still, it is the only way to do things on Windows, so I think this is necessary. I think it's fragile though (for reasons we can't control) so we should only subject ourselves to that fragility when necessary.

Comment (johnwparent) [reply lib/spack/spack/util/compression.py]: Ok that's kind of what I was thinking but wanted to be sure. I'm still hesitant to use something that fragile on Windows. We're only supporting Windows 10 and AFAIK Windows 10 comes with a tar executable that can unpack `.tar`, `.gz`, `.zip` files. Additionally Powershell has a method to do this, or we can write some visual basic wrapped in a `.bat` file that Spack can invoke via `Executable`, so if it's alright with you, I'd prefer to refactor to avoiding `ZipFile` altogether and to explore one of these alternatives.

Comment (scheibelp) [reply lib/spack/spack/util/compression.py]: > if it's alright with you, I'd prefer to refactor to avoid ZipFile altogether and to explore one of these alternatives.

That sounds good to me

**Review by scheibelp — APPROVED** [review]

File: lib/spack/spack/util/compression.py [hunk 4]
```diff
@@ -48,15 +51,14 @@ def _unzip(archive_file):
...
-        from zipfile import ZipFile
-        destination_abspath = os.getcwd()
-        with ZipFile(archive_file, 'r') as zf:
-            zf.extractall(destination_abspath)
-    except ImportError:
-        unzip = which('unzip', required=True)
-        unzip.add_default_arg('-q')
-        return unzip
+    exe = 'unzip'
+    arg = '-xf'
```
Comment (scheibelp) [inline lib/spack/spack/util/compression.py]: Originally `unzip` was using the `-q` option, is this change intentional?

**Review by scheibelp — CHANGES_REQUESTED** [review]
Sorry I was looking over this again and noticed a potential issue (added a question).

I'm guessing this wasn't caught because we don't appear to have any tests for this. That's not specifically a problem with this PR, but can you add a TODO for that (assuming I'm correct that you have to change `arg` for `unzip` anyway)?

**Review by scheibelp — APPROVED** [review]
